packages/packages/store/src/__tests__/store.test.tsx

```diff
@@ -0,0 +1,198 @@
+	}
+}
+
+interface Config {
+	initialValue?: SliceStateRoot['slice']['value'];
+}
+
+const createStoreEntities = ( storeService: ReturnType<typeof createStoreService>, { initialValue = 1 }: Config = {} ) => {
+	const slice = storeService.registerSlice( {
+		name: 'slice',
+		initialState: {
+			value: initialValue,
+		},
+		reducers: {
+			setValue: ( state, action ) => {
+				state.value = action.payload;
+			},
+		},
+	} );
+
+	const store = storeService.createStore();
+
+	const wrapper: React.FC = ( { children } ) => (
+		<StoreProvider store={ store }>
+			{ children }
+		</StoreProvider>
+	);
+
+	return {
+		slice,
+		store,
+		wrapper,
+	};
+};
+
+describe( '@elementor/store', () => {
+	it( 'should verify the initial state of the slice', () => {
+		const storeService = createStoreService();
+
+		const { wrapper } = createStoreEntities( storeService );
+
+		const { result } = renderHook( () => useSelector( ( state: SliceStateRoot ) => state.slice.value ), { wrapper } );
+
+		expect( result.current ).toBe( 1 );
+	} );
+
+	it( 'should update the state value of the slice', () => {
+		const storeService = createStoreService();
+
+		const { slice, wrapper } = createStoreEntities( storeService );
+
+		const { result } = renderHook( () => {
+			const dispatchAction = useDispatch();
+
+			dispatchAction( slice.actions.setValue( 3 ) );
+
+			return useSelector( ( state: SliceStateRoot ) => state.slice.value );
+		}, { wrapper } );
+
+		expect( result.current ).toBe( 3 );
+	} );
+
+	it( 'should throw an error when trying to register a slice with the same name more than once', () => {
+		const storeService = createStoreService();
+
+		storeService.registerSlice( {
+			name: 'slice',
+			initialState: {
+				value: 1,
+			},
+			reducers: {},
+		} );
+
+		expect( () => {
+			storeService.registerSlice( {
+				name: 'slice',
+				initialState: {
+					value: 1,
+				},
+				reducers: {},
+			} );
+		} ).toThrowError();
+	} );
+
+	it( 'should throw an error when trying to re-create the store on the same storeService instance', () => {
+		const storeService = createStoreService();
+
+		storeService.registerSlice( {
+			name: 'slice',
+			initialState: {
+				value: 1,
+			},
+			reducers: {},
+		} );
+
+		storeService.createStore();
+
+		expect( () => storeService.createStore() ).toThrowError();
+	} );
+
+	it( 'should register a middleware that blocks the state update by not running next(action)', () => {
+		const storeService = createStoreService();
+
+		storeService.registerMiddleware( () => () => () => {} );
+
+		const { slice, wrapper } = createStoreEntities( storeService );
+
+		const { result } = renderHook( () => {
+			const dispatchAction = useDispatch();
+
+			dispatchAction( slice.actions.setValue( 4 ) );
+
+			return useSelector( ( state: SliceStateRoot ) => state.slice.value );
+		}, { wrapper } );
+
+		expect( result.current ).toBe( 1 );
+	} );
+
+	it( 'should register a middleware that does not intefer with the state update', () => {
```

Comment:
maybe also test a middleware that changes the state?

Reply:
From what I understand, the main purpose of the middleware is not to change the state value, (when I've tried to change it I got an infinite loop because the state change triggered the same middleware again..) when the middleware needs to change the value you need to start using ReduxThunk and make the change there.. not sure we need it right now. As I see it, we will probably use the middleware to block actions and continue them when required.